web/workflow.R

```diff
@@ -11,184 +11,217 @@
-  settings <- PEcAn.settings::read.settings(
-    file.path(settings$outdir, "pecan.CONFIGS.xml"))
-}
-
-if ((length(which(commandArgs() == "--advanced")) != 0)
-    && (PEcAn.utils::status.check("ADVANCED") == 0)) {
-  PEcAn.utils::status.start("ADVANCED")
-  q();
-}
-
-# Start ecosystem model runs
-if (PEcAn.utils::status.check("MODEL") == 0) {
-  PEcAn.utils::status.start("MODEL")
-  PEcAn.remote::runModule.start.model.runs(settings, stop.on.error = FALSE)
-  PEcAn.utils::status.end()
-}
-
-# Get results of model runs
-if (PEcAn.utils::status.check("OUTPUT") == 0) {
-  PEcAn.utils::status.start("OUTPUT")
-  runModule.get.results(settings)
-  PEcAn.utils::status.end()
-}
-
-# Run ensemble analysis on model output.
-if ("ensemble" %in% names(settings)
-    && PEcAn.utils::status.check("ENSEMBLE") == 0) {
-  PEcAn.utils::status.start("ENSEMBLE")
-  runModule.run.ensemble.analysis(settings, TRUE)
-  PEcAn.utils::status.end()
-}
-
-# Run sensitivity analysis and variance decomposition on model output
-if ("sensitivity.analysis" %in% names(settings)
-    && PEcAn.utils::status.check("SENSITIVITY") == 0) {
-  PEcAn.utils::status.start("SENSITIVITY")
-  runModule.run.sensitivity.analysis(settings)
-  PEcAn.utils::status.end()
-}
-
-# Run parameter data assimilation
-if ("assim.batch" %in% names(settings)) {
-  if (PEcAn.utils::status.check("PDA") == 0) {
-    PEcAn.utils::status.start("PDA")
-    settings <- PEcAn.assim.batch::runModule.assim.batch(settings)
-    PEcAn.utils::status.end()
-  }
-}
-
-# Run state data assimilation
-if ("state.data.assimilation" %in% names(settings)) {
-  if (PEcAn.utils::status.check("SDA") == 0) {
-    PEcAn.utils::status.start("SDA")
-    settings <- sda.enfk(settings)
-    PEcAn.utils::status.end()
-  }
-}
-
-# Run benchmarking
-if ("benchmarking" %in% names(settings)
-    && "benchmark" %in% names(settings$benchmarking)) {
-  PEcAn.utils::status.start("BENCHMARKING")
-  results <- papply(settings, function(x) calc_benchmark(x, bety))
-  PEcAn.utils::status.end()
-}
-
-# Pecan workflow complete
-if (PEcAn.utils::status.check("FINISHED") == 0) {
-  PEcAn.utils::status.start("FINISHED")
-  PEcAn.remote::kill.tunnel(settings)
-  db.query(paste("UPDATE workflows SET finished_at=NOW() WHERE id=",
-                 settings$workflow$id, "AND finished_at IS NULL"),
-           params = settings$database$bety)
-
-  # Send email if configured
-  if (!is.null(settings$email)
-      && !is.null(settings$email$to)
-      && (settings$email$to != "")) {
-    sendmail(settings$email$from, settings$email$to,
-             paste0("Workflow has finished executing at ", base::date()),
-             paste0("You can find the results on ", settings$email$url))
-  }
-  PEcAn.utils::status.end()
-}
-
-db.print.connections()
-print("---------- PEcAn Workflow Complete ----------")
+suppressMessages(library("PEcAn.all"))
+suppressMessages(library("PEcAn.utils"))
+suppressMessages(library("RCurl"))
+suppressMessages(library("optparse"))
```

Comment:
this is a package that needs to to be installed first. I'm tempted to suggest adding it to base/utils since that is also where the status functions are.

Reply:
@robkooper did we ever document when a package should be added as 'Imports' vs 'Suggests'?

Reply:
https://pecanproject.github.io/pecan-documentation/develop/package-dependencies.html

Reply:
Since the parsing is neatly encapsulated already, I suggest moving the definition of `get_args` into a package (utils or workflow seem like the obvious candidates) and calling it from there — then folks reading the workflow file just see eg `args <- PEcAn.utils::get_args(commandArgs())` and don’t have to grapple with the full function definition

Reply:
I'm not sure I entirely follow this, but I think I disagree. The code in `get_args()` could easily live in the main body of the program (or the `main()` that I wrote), but it needs to be a part of this program. It's not a generic function. It describes and validates the parameters to this program only.

FWIW, I'm cargo-culting my ideas from Python. I've written extensively on these ideas (see Tiny Python Projects), and Kristina saw me present on the first chapter where I spell out these ideas of encapsulating ideas into functions which hopefully can be tested, documenting command-line arguments, running integration tests for a program, etc. Many Perl/Python/R programs write all the code in the "main" namespace and don't bother with functions. This has the effect, e.g., of making all the variables global to the program which is not a best practice. I try to write small (<50 lines), *pure* (no global/environmental variables, hidden state, etc.) functions (and unit tests!) to try to break down my programs into understandable functions that can be composed.

Every command-line program I write in Python starts in a `main()` function and almost always has a `get_args()` function for handling the command-line arguments. I have written a Python program called `new.py` to create new programs with these functions:

https://github.com/kyclark/tiny_python_projects/blob/master/bin/new.py

The idea is to bring a standard interface to my programs and to adhere to the Python communities best practices. The "argparse" module is standard and widely used. Most programmers (esp coming from C languages, Java, etc.) would immediately understand starting in a `main()` function. Small functions tend to limit scoping of variables and make programs easier to understand AND TEST.

I feel strongly that every command-line program ought to use the same interface for documenting and validating the arguments. R's offerings in this space seem limited to "optparse" which I have demonstrated here. It's lacking in many of the features from Python's "argparse" such as automatically validating that an argument is a readable text file so you note that I had to check this manually. There appears to be an R package called "argparse" which in some way seems to rely on the existence of Python to handle the argument processing, so I'm not sure if this is a better or worse option.

In addition to validating the arguments, every command-line program ought to generate a "usage" statement when run with the standard "-h" and "--help" flags. The only reason I found out about the "--continue" flag was because I read the source code, which is an unreasonable expectation for the user/programmer. Additionally, it turns out that the "read.settings()` function ALSO parses the `commandArgs()` which seems highly unusual and undesirable as it could create other hidden, undocumented options for the program. 

Only by manually tracing and reading all the source code could a person ever know all the options that the code accepts. This is not a best practice. ALL of the options should be explicitly stated in documentation/usage and should be VALIDATED IMMEDIATELY upon program execution. Every program should strive to fail as early as possible, e.g., if some file is required then verify both that the argument is present and is actually a readable file. 

I would be happy to demonstrate my ideas using Python and trying to find reasonable parallels in R. I could try to explain unit and integration tests, and point out some of the flaws in dynamically typed languages like R/Python that can easily lead to unmaintainable code. I could explain some of the errors that I've seen in the current codebase with things like scoping of variables, impure and extremely long functions with far too many parameters, fatal errors that are ignored, lack of exceptions and traceback information that seems to make debugging R a particular bear. 

In short, I would recommend adding "optparse" as a standard module for the project and encouraging every command-line program to adopt this. It's not necessary to have `main()` and `get_args()` functions, but just defining the program parameters can easily take up 100+ lines for a complicated program, so it could make the main body of the code easier to read if it's a separate function.

Reply:
I think I'm going to concur with @infotroph here. Definitely like the `get_args` function. Agree that it sees like it should go in a package, such as `settings`, which likewise only makes sense in the context of PEcAn's workflow. Agree that I'd prefer to not put the workflow in a `main()`. Agree that I don't want to suppress warnings. Would also add a request that command line args be documented in the model documentation as part of this PR.

Reply:
> It's not a generic function. It describes and validates the parameters to this program only.

This is a very good point that I wasn't thinking about! I completely understand the philosophical objection, but pragmatically there's plenty of precedent in PEcAn for choosing encapsulation over inline usage even when it's not fully generic -- c.f. [`PEcAn.DB::betyConnect`](https://github.com/PecanProject/pecan/blob/develop/base/db/R/query.dplyr.R#L5), which hard-codes a very specific (and arguably Rube-Goldberg, but still useful!) set of database parameters.

Let's discuss your other points below and keep the line comments focused on how to handle this specific line of code.